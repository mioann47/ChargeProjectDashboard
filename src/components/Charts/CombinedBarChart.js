
import React, { Component } from 'react';
import { ResponsiveBar } from '@nivo/bar'
import PropTypes from 'prop-types';
import { ShowAt, HideAt } from 'react-with-breakpoints';

import csvPng from '../../images/csv.png'
import { CSVLink } from "react-csv";
export default class CombinedBarChart extends Component {


  render() {
    return (<>

      <div style={{ textAlign: 'center', justifyContent: 'center' }}>
      <h3>{this.props.title} {!this.props.hideCSV&&<CSVLink
          filename={this.props.title.replace(/ /g,'').replace(/\//g,'')+".csv"}
            data={this.props.data}
          >
            <img width={15} height={15} src={csvPng} alt="Click to download chart data" />
          </CSVLink>}</h3></div>
      <div style={{ height: this.props.height }}>
        <HideAt breakpoint="small">
          <ResponsiveBar
          borderWidth={1}
            labelFormat={d => <tspan y={-8} x={30} font-weight="bold" font-size='13px'>{d}</tspan>}
            data={this.props.data}
            margin={{ top: 50, right: this.props.legend ? 150 : 5, bottom: 65, left: 60 }}
            padding={this.props.data.length>4?0.3:0.8}
            keys={this.props.keys}
            indexBy={this.props.indexBy}
            groupMode="grouped"
            colors={this.props.colors}
            colorBy="id"
            borderColor={{ from: 'color', modifiers: [['darker', 2]] }}
            axisTop={null}
            axisRight={null}
            axisBottom={{
              tickSize: 5,
              tickPadding: 5,
              tickRotation: -45,
              legend: this.props.legendBottom,
              legendPosition: 'middle',
              legendOffset: 55
            }}

            axisLeft={{
              tickSize: 5,
              tickPadding: 5,
              tickRotation: 0,
              legend: this.props.legendLeft,
              legendPosition: 'middle',
              legendOffset: -52
            }}
            labelSkipWidth={30}
            labelSkipHeight={30}
            labelTextColor={{ from: 'color', modifiers: [['darker', 4]] }}
            legends={this.props.legend === true ? [
              {
                dataFrom: 'keys',
                anchor: 'bottom-right',
                direction: 'column',
                justify: false,
                translateX: 120,
                translateY: 0,
                itemsSpacing: 2,
                itemWidth: 100,
                itemHeight: 20,
                itemDirection: 'left-to-right',
                itemOpacity: 0.85,
                symbolSize: 20,
                effects: [
                  {
                    on: 'hover',
                    style: {
                      itemOpacity: 1
                    }
                  }
                ]
              }
            ] : []}
            animate={true}
            motionStiffness={90}
            motionDamping={15}
          />
        </HideAt>
        <HideAt breakpoint="mediumAndAbove">
          <ResponsiveBar
          borderWidth={1}
            data={this.props.data}
            margin={{ top: 50, right: 5, bottom: 65, left: 60 }}
            padding={this.props.data.length>4?0.3:0.8}
            keys={this.props.keys}
            indexBy={this.props.indexBy}
            groupMode="grouped"
            colors={this.props.colors}
            colorBy="id"
            borderColor={{ from: 'color', modifiers: [['darker', 2]] }}
            axisTop={null}
            axisRight={null}
            axisBottom={{
              tickSize: 5,
              tickPadding: 5,
              tickRotation: -45,
              legend: this.props.legendBottom,
              legendPosition: 'middle',
              legendOffset: 55
            }}

            axisLeft={{
              tickSize: 5,
              tickPadding: 5,
              tickRotation: 0,
              legend: this.props.legendLeft,
              legendPosition: 'middle',
              legendOffset: -52
            }}
            labelSkipWidth={30}
            labelSkipHeight={30}
            labelTextColor={{ from: 'color', modifiers: [['darker', 4]] }}
            labelFormat={d => <tspan y={-8}  x={30} font-weight="bold" font-size='13px'>{d}</tspan>}
            animate={true}
            motionStiffness={90}
            motionDamping={15}
          />
        </HideAt>
      </div></>)
  }


}

CombinedBarChart.propTypes = {
  title: PropTypes.string,
  data: PropTypes.array.isRequired,
  legendBottom: PropTypes.string.isRequired,
  legendLeft: PropTypes.string.isRequired,
  colors: PropTypes.oneOfType([PropTypes.array, PropTypes.object, PropTypes.func]).isRequired,
  legend: PropTypes.bool.isRequired,
  keys: PropTypes.array.isRequired,
  indexBy: PropTypes.string.isRequired
};