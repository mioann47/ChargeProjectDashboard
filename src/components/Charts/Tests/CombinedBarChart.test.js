import React from 'react'
import { mount,shallow } from 'enzyme'
import renderer from 'react-test-renderer';
import CombinedBarChart from '../CombinedBarChart'
const data = [
{"_id":{"index":1},"data1":20.68,"data2":12.49,"num":1,"index":"January"},
{"_id":{"index":2},"data1":18.67,"data2":11.13,"num":2,"index":"February"},
{"_id":{"index":3},"data1":20.73,"data2":12.55,"num":3,"index":"March"},
{"_id":{"index":4},"data1":20.04,"data2":12.06,"num":4,"index":"April"},
{"_id":{"index":5},"data1":20.67,"data2":12.36,"num":5,"index":"May"},
{"_id":{"index":6},"data1":20.01,"data2":12.07,"num":6,"index":"June"},
{"_id":{"index":7},"data1":20.65,"data2":12.35,"num":7,"index":"July"},
{"_id":{"index":8},"data1":20.67,"data2":12.5,"num":8,"index":"August"},
{"_id":{"index":9},"data1":20.02,"data2":12,"num":9,"index":"September"},
{"_id":{"index":10},"data1":20.69,"data2":12.44,"num":10,"index":"October"},
{"_id":{"index":11},"data1":19.97,"data2":11.93,"num":11,"index":"November"},
{"_id":{"index":12},"data1":20.67,"data2":12.49,"num":12,"index":"December"}]

it('should render bar chart correctly', () => {

    const component = renderer.create(
        <CombinedBarChart title={'Chart Title'}
        indexBy="index"
        keys={[ "data1","data2" ]}
        hideCSV={true}
        data={data}
        height={300}
        legendBottom={'BottomLegend'}
        legendLeft={'LeftLegend'}
        colors={['#a6761d','#e7298a','#66a61e','#666666','#7570b3','#e6ab02','#d95f02','#1b9e77']}
        legend={false}
        labelTooltip={'tooltipLavel'} />

    )

let tree = component.toJSON()
expect(tree).toMatchSnapshot()
})



it('will check props', () => {
    const props = mount( <CombinedBarChart title={'Chart Title'}
    keys={[ "data1","data2" ]}
    indexBy="index"
    hideCSV={true}
    data={data}
    height={300}
    legendBottom={'BottomLegend'}
    legendLeft={'LeftLegend'}
    colors={['#a6761d','#e7298a','#66a61e','#666666','#7570b3','#e6ab02','#d95f02','#1b9e77']}
    legend={false}
    labelTooltip={'tooltipLavel'} />).props();
    expect(props.indexBy).toEqual("index");
    expect(props.keys).toEqual([ "data1","data2" ]);
    expect(props.height).toEqual(300);
    expect(props.data).toEqual(data);
    expect(props.legendBottom).toEqual('BottomLegend');
    expect(props.legendLeft).toEqual('LeftLegend');
    expect(props.colors).toEqual(['#a6761d','#e7298a','#66a61e','#666666','#7570b3','#e6ab02','#d95f02','#1b9e77']);
    expect(props.labelTooltip).toEqual('tooltipLavel');
    expect(props.legend).toEqual(false);
  });
  