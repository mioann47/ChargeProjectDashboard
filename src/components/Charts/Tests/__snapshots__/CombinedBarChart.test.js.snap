// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render bar chart correctly 1`] = `
Array [
  <div
    style={
      Object {
        "justifyContent": "center",
        "textAlign": "center",
      }
    }
  >
    <h3>
      Chart Title
       
    </h3>
  </div>,
  <div
    style={
      Object {
        "height": 300,
      }
    }
  />,
]
`;
